_books/ion-1-1/src/security-considerations.md

```diff
@@ -0,0 +1,156 @@
+Ion 1.1 is primarily susceptible to denial-of-service (DoS) attacks that attempt to cause an error condition in the receiving
+system or consume excessive system resources. As with many such attacks, the strongest defense is to not accept any 
+untrusted input, but that defense is not always compatible with the business requirements of the receiving application.
+
+This document addresses various types of attacks, assuming that it is not possible to avoid accepting untrusted input.
+
+_The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in RFC 2119._
+
+## Data expansion denial-of-service
+
+An attacker could craft an input that is relatively small, but upon expansion, produces something thousands or millions
+of times larger.
+
+For many use cases, the expansion of a template macro will grow linearly with the size of its input. However, it is 
+possible to create macros with expansions that grow at greater rates. Using [`for`](macros/special_forms.md#for) we can 
+nest an arbitrary number of loops to create a macro expansion with a polynomial growth rate. Using the 
+[`repeat`](macros/system_macros.md#repeat) macro, we can create classes of inputs with expansions that grow 
+exponentially in relation to the input.
+
+For example, this input is less than 250 characters when encoded as Ion text (and omitting all optional whitespace). 
+In Ion binary, it requires only 74 bytes. For each additional level of nesting, only 20 additional characters (text) or
+6 additional bytes (binary) are required, but it increases the number of expanded values by 2147483647 times.
+```ion
+$ion_1_1
+(:repeat 2147483647
+  (:repeat 2147483647
+    (:repeat 2147483647
+      (:repeat 2147483647
+        (:repeat 2147483647
+          (:repeat 2147483647
+            (:repeat 2147483647
+              (:repeat 2147483647
+                (:repeat 2147483647
+                  (:repeat 2147483647
+                    (:repeat 2147483647 "abc")))))))))))
+```
+The expansion of these e-expressions results in a stream of ~450 [googol](https://en.wikipedia.org/wiki/Googol) string
+values. Any attempt to hold all of this in memory or write it to disk will exhaust all available resources and 
+eventually fail. Even an attempt to count the length of the stream, while it may theoretically succeed if using an 
+appropriate `BigInteger` type, will require a considerable amount of CPU operations (over a googol), and even the 
+fastest processors will require many millennia to completely count the number of values in the stream.
+
+Even without using `repeat` or `for`, a [Billion laughs attack](https://en.wikipedia.org/wiki/Billion_laughs_attack)
+could exist for any data format with macro expansion, and it is certainly possible with Ion 1.1.
```

Comment:
Neat, we've discussed this attack a bunch of times but I didn't know it had an established name.